llvmPackages_*.clang-unwrapped: make scan-build work

sternenseemann opened the pull request:
* scan-build and c*-analyzer are written in Perl, so we need to add it
  as a buildInput, so the interpreter is patched.

* For scan-build to discover c*-analyzer, the `bin` and `libexec`
  directories need to be in the same prefix. Since this is no big
  trouble for clang, we can just move the libexec directory from the
  `lib` to the `out` output.

To be tested how this interacts with bootstrapping for macOS and perhaps `pkgsLLVM`.
Additional perl dependency may be a problem in some cases?

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

sternenseemann committed:
llvmPackages_*.clang-unwrapped: make scan-build-py work

Move scan-build-py and its dependencies libear, libscanbuild,
analyze{c++,cc} and intercept-{cc,c++} where they are patchShebanged
without incurring an extra dependency on python for $out.

libear, libscanbuild, libexec/{intercept,analyze}-c* are tracked in
clang/tools/scan-build-py, so I assume they aren't used by anything
else.

schuelermine commented:
Checked build & it works

sternenseemann commented:
Unsure whether this creates a problem for darwin in its current form since perl is not in bootstrap-tools I believe?

sternenseemann commented:
Can I get some feedback on this (again)? For me there are the following questions:

* Should we move `scan-build` to `dev` instead of `bin`?
* What about the closure size with `perl` included? Would it be better to require the user of `scan-build` to “bring their own perl” and leave the `/usr/bin/env perl` shebang unpatched?

alyssais commented:
Haven't had much of a look, but please no unpatched shebangs — it would make it not run in the Nix sandbox, where there's no /usr/bin/env.

RossComputerGuy commented:
This PR has merge conflicts, has gotten stale, and has CI failures, is this PR still something we want to have done?

sternenseemann commented:
Yes, we probably should. It's still an issue that scan-build doesn't work.

RossComputerGuy commented:
Alright, any plans on rebasing the PR to resolve the conflicts?

cloudripper commented:
Are there any plans to move forward with this PR? This is still an open issue.

Bradley-Cooley commented:
@sternenseemann I'd love to see this land - not a contributor, but anything I can do to help?

RossComputerGuy commented:
> but anything I can do to help?

Unfortunately not, this PR requires reading which @sternenseemann has to do.

sternenseemann committed:
llvmPackages_*.clang-unwrapped: make scan-build work if perl in PATH

This is a conservative take on making scan-build work. It basically
needs two things

- perl, idelly via a patched shebang
- Access to {ccc,c++}-analyzer which is provided easily by moving
  libexec. Patching scan-build to look into $lib would defeat the
  purpose of the separate outputs, of course.

The former is not provided via this change, instead /usr/bin/env and
perl in PATH are required. This is mainly to avoid the closure size
increase of ~100MB caused by perl plus potential bootstrapping trouble
on darwin. (scan-build-py can be made to work without much less problems
since clang-unwrapped.python already depends on python.) We can probably
improve on this by adding scan-build* to clang-tools.

sternenseemann commented:
Redid the entire change on current nixpkgs and also made `scan-build-py` work (which is trivial). The change for `scan-build` is more conservative now and relies on an impure `perl` to avoid closure size concerns. I think we should probably integrate scan-build into clang-tools in the future to solve this. (Or add a `perl` output to clang, but ugh…)

sternenseemann commented:
@RossComputerGuy Just found this again… Do we still want to do this?

RossComputerGuy commented:
> Do we still want to do this?

If we want the tools still, sure
